File: 121-best-time-to-buy-and-sell.py
# look at JS file for reasoning and explanation

class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        buy = 0
        sell = 1
        max_profit = 0

        while sell < len(prices):
            curr_profit = prices[sell] - prices[buy]
            if curr_profit > 0:
                max_profit = max(max_profit, curr_profit)
            else:
                buy = sell
            sell += 1

        return max_profit

#  faster for loop implementation?

class Solution(object):
    def maxProfit(self, prices):
        max_profit = 0
        buy = prices[0]

        # [7,1,5,3,6,4]
        for price in prices:
            if price < buy:
                buy = price
            if price - buy > max_profit:
                max_profit = price- buy

        return max_profit

File: 121-best-time-to-buy-and-sell.js
/*
given array prices where prices[i] is the price of a stock on the ith day
return the max profit you can achieve from prices
return 0 if no profit can be made
*/

/*
use 2 pointers (buy=0, sell=1) for sliding window:
1.) have a profit variable to track max profit
2.) iterate across prices with a while loop (while right pointer is less than length of prices arr)
3.) calculate current profit based on pointers indices
4.) if positive profit, compare against max profit and update if current is greater
5.) if negative profit, update buy pointer to the sell pointer
6.) increment sell pointer at the end of the loop
7.) return max profit outside the loop
*/

// time: O(n), n is number of days in prices
// space: O(n), only 4 variables used

var maxProfit = function(prices) {
    let buy = 0;
    let sell = 1;
    let maxProfit = 0;

    while (sell < prices.length) {
        let currProfit = prices[sell] - prices[buy];

        if (currProfit > 0) {
            if (currProfit > maxProfit) maxProfit = currProfit;
        } else {
            buy = sell;
        }

        sell++;
    }

    return maxProfit;
};
